arch/arm/src/stm32h7/stm32_fmc.c

```diff
@@ -39,40 +39,387 @@
+
+/* The bits in FMC_SDTR we will alter at initialization */
+#define FMC_SDTR_MASK           (FMC_SDTR_TMRD_MASK | FMC_SDTR_TXSR_MASK | \
+    FMC_SDTR_TRAS_MASK | FMC_SDTR_TRC_MASK | FMC_SDTR_TWR_MASK | FMC_SDTR_TRP_MASK | \
+    FMC_SDTR_TRCD_MASK)
+
+/* The timeout while waiting for SDRAM controller to initialize, in us */
+#define SDRAM_INIT_TIMEOUT      (1000)
+
+
+/****************************************************************************
+ * Private data
+ ****************************************************************************/
+
+static uint32_t fmc_gpios[] =
+{
+    BOARD_FMC_GPIO_CONFIGS
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static void stm32_fmc_sdram_init(void);
+static int stm32_fmc_sdram_wait(unsigned timeout);
+
 
 /****************************************************************************
  * Public Functions
  ****************************************************************************/
 
 /****************************************************************************
- * Name: stm32_fmc_enable
+ * Name: stm32_fmc_init
+ *
+ * Description:
+ *   Initialize the FMC peripherial. Because FMC initialization is highly
+ *   dependent on the used parts, definition of the initial values for FMC
+ *   registers is mostly left to board designer.
+ *
+ *   Typically called from up_addregion().
+ *
+ ****************************************************************************/
+
+void stm32_fmc_init(void)
+{
+  uint32_t regval;
+
+  /* Reset the FMC on the AHB3 bus */
+
+  regval  = getreg32(STM32_RCC_AHB3RSTR);
+  regval |= RCC_AHB3RSTR_FMCRST;
+  putreg32(regval, STM32_RCC_AHB3RSTR);
+
+  /* Leave reset state */
+
+  regval &= ~RCC_AHB3RSTR_FMCRST;
+  putreg32(regval, STM32_RCC_AHB3RSTR);
+
+  /* Set FMC clocking */
+
+  modreg32 (BOARD_FMC_CLK, RCC_D1CCIPR_FMCSEL_MASK, STM32_RCC_D1CCIPR);
+
+  /* Set up FMC GPIOs */
+
+  for (regval = 0; regval < ARRAY_SIZE(fmc_gpios); regval++)
+    stm32_configgpio(fmc_gpios[regval]);
+
+  /* Set up FMC registers */
+
+#ifdef BOARD_FMC_SDCR1
+  modreg32(BOARD_FMC_SDCR1, FMC_SDCR_MASK, STM32_FMC_SDCR1);
+#endif
+#ifdef BOARD_FMC_SDCR2
+  modreg32(BOARD_FMC_SDCR2, FMC_SDCR_MASK, STM32_FMC_SDCR2);
+#endif
+
+#ifdef BOARD_FMC_SDTR1
+  modreg32(BOARD_FMC_SDTR1, FMC_SDTR_MASK, STM32_FMC_SDTR1);
+#endif
+#ifdef BOARD_FMC_SDTR2
+  modreg32(BOARD_FMC_SDTR2, FMC_SDTR_MASK, STM32_FMC_SDTR2);
+#endif
+
+  /* Enable the FMC peripherial */
+
+  modreg32(FMC_BCR_FMCEN, FMC_BCR_FMCEN, STM32_FMC_BCR1);
+
+  /* Initialize the SDRAM chips themselves */
+
+#ifdef HAVE_FMC_SDRAM
+  stm32_fmc_sdram_init();
+#endif
+
+  /* Set up SDRAM refresh timings */
+
+#ifdef BOARD_FMC_SDRAM_REFR_PERIOD
+  /* ... The programmed COUNT value must not be equal to the sum of the following timings:
+     TWR+TRP+TRC+TRCD+4 memory clock cycles */
```

Comment:
*/ must be on a separate line per coding standard
All comments must be followed by a blank line per coding standard.